### PR TITLE
Json serialization persistor

### PR DESCRIPTION
**Pull request by kjanosz**
Commits are not squashed yet because this PR is mainly for discussion purpose.

### CODE REVIEW

File: src/main/scala/org/scalameter/persistence/codecs/domain.scala [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  def decode(jV: JValue)(implicit hint: TypeHint): Try[History] = Try {
+    jV match {
+      case jObj: JObject => History(
+        results = jObj.get("results").asInstanceOf[JArray].vs.map(ResultCodec.unsafeDecode),
+        infomap = KeyMapCodec.unsafeDecode(jObj.get("infomap"))
+      )
+      case _ => sys.error(s"Expected JObject in HistoryCodec. Got: $jV.")
+    }
+  }
+}
```
Comment (kjanosz) [inline src/main/scala/org/scalameter/persistence/codecs/domain.scala]: I think that all this codecs can be replaced by some reflection, or better, macro based generic codec. Do you think is it worth?

File: src/main/scala/org/scalameter/persistence/codecs/complex.scala [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        case v => sys.error(s"Expected JObject as array element in SeqCodec. Got: $v.")
+      }(collection.breakOut)
+      case _ => sys.error(s"Expected JArray in SeqCodec. Got: $jV.")
+    }
+  }
+}
+
+object KeyMapCodec extends JSONCodec[Map[Key[_], Any]] {
+  def encode(obj: Map[Key[_], Any]): JValue = JArray.fromSeq(obj.map { case (k, v) =>
+    KeyValueCodec.encode(k.fullname -> v)
```
Comment (kjanosz) [inline src/main/scala/org/scalameter/persistence/codecs/complex.scala]: @axel22: Saving of type hints here can be avoided, but it would require to enforce TypeTag upon key creation, like this:
`class Key[T: scala.reflect.runtime.universe.TypeTag]`
and to add type reconstruction based on it.
Do you see any contraindications?

File: src/main/scala/org/scalameter/persistence/codecs/domain.scala (hunk 1 above)
Comment (axel22) [inline src/main/scala/org/scalameter/persistence/codecs/domain.scala]: It seems like: if you add a new field to some class,
you might forget to update the `decode` method for it. Fortunately, most of our classes have all the fields in the constructor, so that would be detected during compilation, so I don't think it is a huge problem.

One bigger potential issue is that the codecs write potentially the same objects both when they write to JSON, and recreate them when they write back. If the `Context`, which could take a lot of space, is duplicated for each curve (and curves typically share the same `Context`  object), it takes up a lot of space. Perhaps it would be useful to cache some of these objects during a codec run.

File: src/main/scala/org/scalameter/persistence/codecs/complex.scala (hunk 2 above)
Comment (axel22) [inline src/main/scala/org/scalameter/persistence/codecs/complex.scala]: I don't think adding the `TypeTag` to `T` in `Key` should cause any problems - feel free to do this.

However, you might need to store the implicit argument as a field in `Key` to be able to reuse it.

**Comment by kjanosz** [comment]
I've updated PR a little bit - now json persistor uses Jackson under the hood. 
There is a one problem though:
`axisData: ListMap[String, Any]` in `Parameters` is serialized without type hints, so it's deserialized to most simple format, e.g. map is used instead of any class that was before serialization, ints are used instead of longs if serialized number is small enough, etc.
Adding type hints is a little problematic. 
Using techniques described at [Jackson wiki](https://github.com/FasterXML/jackson-docs/wiki/JacksonPolymorphicDeserialization) doesn't solve it.
Firstly, there are some problem during serialization:
- adding them by `mapper.enableDefaultTyping(...)` at global level messes up other parts of json 
- adding them directly by using `@JsonTypeInfo` works only on first level, e.g. given `List[Long](1, 2, 3)` only type hint with list type is added, there are no type hints for values inside list.

Secondly, I haven't managed to deserialize json with type hints despite playing with settings a little bit - there are either errors  or hints are ignored and e.g. parsed as `axisData` map values. 

The only solution I currently see is to use Jackson inside `axisData` only for serializing/deserializing primitives and field names whilst saving own typing hints and reading them and instating all complex objects recursively using scala reflection. 

@axel22: What do you think about it?

File: scalameter-core/src/main/scala/org/scalameter/Parameter.scala [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    val i = str.lastIndexOf('|')
+    if (i == -1) sys.error("""Invalid parameter string. It should have following form "name|manifest".""")
+    val key = str.substring(0, i)
+    val manifestString = str.substring(i + 1)
+    val parser = new ManifestParser
+    val m = parser.parse(parser.manifest, new parser.PackratReader(new CharSequenceReader(manifestString))).getOrElse(
+      sys.error(s"""Invalid manifest string: "$manifestString".""")
+    )
+    Parameter(key)(m)
+  }
```
Comment (axel22) [inline scalameter-core/src/main/scala/org/scalameter/Parameter.scala]: I took a look at these parsers, and they seem to do the right thing.

I appreciate the effort you invested into this design, however, I'm not sure if we want to go in this direction. I was really hoping that we could have some serializer that just works out-of-the-box. Here, it looks like we're trying to reimplement the wheel by creating a custom encoding of all the Scala types.
The costs are:
- increased complexity
- increased space consumption
- increased maintenance costs

At the same time, the benefits of supporting any type whatsoever in the Parameters is questionable - the provided basic `Gen` objects mostly create primitive values. Those basic `Gen` objects can compose into more complex data types, but they always need to serialize only the values from the basic `Gen`s.

There are (at most) 1% of users that write custom `Gen` implementations. Of those, hardly anybody uses types other than the primitives and `String`. It seems like it is not worth to add such a complexity only to appease such users - if they insist on custom generators, these users can still use the `SerializationPersistor`.

Here's what I would suggest instead:
- introduce (to the `core` project) a (type-)class `Pickler[T]`:

        abstract class Pickler[T] {
          def pickle(x: T): Array[Byte] // byte array is general enough - serializers can use base64 to escape the world of bytes
          def unpickle(a: Array[Byte]): T
        }

- implement `Pickler`s for the basic data types - primitives and `String`
- ensure that `Gen` factories (especially, `single` and `enumerate`) take the `Pickler[T]` typeclass
- enrich `Parameters` with the `Pickler` arguments - you could replace `Any` in `axisData` with just `(Any, Pickler[Any])` - it is fine if we're internally not always completely type-safe

In the main project:
- introduce your JSON serializer that will encode the `Pickler` classname along with the array contents

We can do something similar if we need to persist `Key`s, but first focus on the `Parameters`.

This way:
- people get a binary-compatible serializer that works with basic data-types
- people who really want more can do their own picklers - we don't need to support exotic picklers for them - or they can revert to `SerializationPersistor`

File: project/Build.scala [hunk 4]
```diff
@@ -167,10 +167,12 @@ object ScalaMeterBuild extends MechaRepoBuild {
       "org.apache.commons" % "commons-math3" % "3.2",
       "org.scala-lang" % "scala-reflect" % "2.11.0",
       "org.scala-lang.modules" %% "scala-xml" % "1.0.1",
-      "org.scala-lang.modules" %% "scala-parser-combinators" % "1.0.1"
+      "org.scala-lang.modules" %% "scala-parser-combinators" % "1.0.1",
+      "com.fasterxml.jackson.module" %% "jackson-module-scala" % "2.5.2"
       )
     case Some((2,10)) => List (
-      "org.apache.commons" % "commons-math3" % "3.2"
+      "org.apache.commons" % "commons-math3" % "3.2",
```
Comment (axel22) [inline project/Build.scala]: Same here.

File: project/Build.scala [hunk 5]
```diff
@@ -167,10 +167,12 @@ object ScalaMeterBuild extends MechaRepoBuild {
       "org.apache.commons" % "commons-math3" % "3.2",
       "org.scala-lang" % "scala-reflect" % "2.11.0",
       "org.scala-lang.modules" %% "scala-xml" % "1.0.1",
-      "org.scala-lang.modules" %% "scala-parser-combinators" % "1.0.1"
+      "org.scala-lang.modules" %% "scala-parser-combinators" % "1.0.1",
```
Comment (axel22) [inline project/Build.scala]: Since persistors are not defined in the core package, we should move this to the dependencies of the main project.

**Comment by axel22** [comment]
Nice work! I left a bunch of comments in the diffs.
Btw, please try to make your commits more incremental in the future. Even if you don't yet have a fully implemented feature - seeing your work in progress is helpful!

We can discuss this once you are available again - *don't* let this take any time until your exams are over (and good luck!)

File: scalameter-core/src/main/scala/org/scalameter/json/package.scala [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package org.scalameter
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+import com.fasterxml.jackson.module.scala.experimental.ScalaObjectMapper
+
+
+package object json {
```
Comment (axel22) [inline scalameter-core/src/main/scala/org/scalameter/json/package.scala]: Please move this to the main project.

File: src/main/scala/org/scalameter/Parameters.scala [hunk 7]
```diff
@@ -1,20 +1,31 @@
...
+    }
+    copy(axisData = newAxisData.asInstanceOf[immutable.ListMap[Parameter[_], Any]])
+  }
 
   override def toString = s"Parameters(${axisData.map(t => t._1 + " -> " + t._2).mkString(", ")})"
 }
 
 object Parameters {
-  def apply(xs: (String, Any)*) = new Parameters(immutable.ListMap(xs: _*))
+  private type P = T forSome { type T }
```
Comment (axel22) [inline src/main/scala/org/scalameter/Parameters.scala]: Please, avoid using existential types. We want our code robust against future possible changes, and the benefit from having them here is not high.

File: src/main/scala/org/scalameter/Parameters.scala [hunk 8]
```diff
@@ -1,20 +1,31 @@
 package org.scalameter
 
+import scala.language.existentials
 import scala.collection.{Iterable, immutable}
+import Parameters.P
+
 
 @SerialVersionUID(4203959258570851398L)
-case class Parameters(axisData: immutable.ListMap[String, Any]) {
+case class Parameters(axisData: immutable.ListMap[Parameter[P], P]) {
```
Comment (axel22) [inline src/main/scala/org/scalameter/Parameters.scala]: It is unconventional to use single letter abstract type names for concrete types. See my comment about existentials below.

File: src/main/scala/org/scalameter/Parameters.scala [hunk 9]
```diff
@@ -1,20 +1,31 @@
...
   def ++(that: Parameters) = Parameters(this.axisData ++ that.axisData)
-  def apply[T](key: String) = axisData.apply(key).asInstanceOf[T]
-  def map(f: ((String, Any)) => (String, Any)) = Parameters(axisData.map(f))
+  def apply[T: Manifest](key: String) = axisData.apply(Parameter[T](key)).asInstanceOf[T]
```
Comment (axel22) [inline src/main/scala/org/scalameter/Parameters.scala]: I suggest you keep it simple, and make `Parameters` a map from `String` to `(Any, Pickler[Any])`.

File: src/main/scala/org/scalameter/PerformanceTest.scala [hunk 10]
```diff
@@ -91,7 +91,7 @@ object PerformanceTest {
    */
   trait HTMLReport extends PerformanceTest {
     import reporting._
-    def persistor: Persistor = new persistence.SerializationPersistor
+    def persistor: Persistor = new persistence.GZIPJSONSerializationPersistor
```
Comment (axel22) [inline src/main/scala/org/scalameter/PerformanceTest.scala]: The consequence of doing this is potentially breaking existing tests out there.
When we get to this point, please do this change in a separate commit - it might be necessary to introduce a new test template, or provide a migration tool from the old formats (the first is probably simpler).

File: src/main/scala/org/scalameter/japi/GenFactory.scala [hunk 11]
```diff
@@ -33,8 +33,8 @@ class VoidGen(axisName: String) extends JavaGenerator[Void] {
 }
 
 
-class SingleGen[T](axisName: String, v: T) extends JavaGenerator[T] {
-  def get = Gen.single(axisName)(v)
+class SingleGen[T](axisName: String, v: T, manifest: Manifest[T]) extends JavaGenerator[T] {
```
Comment (axel22) [inline src/main/scala/org/scalameter/japi/GenFactory.scala]: I'd embed a `Pickler[T]` instead of the `Manifest[T]` here.

File: src/main/scala/org/scalameter/japi/GenFactory.scala [hunk 12]
```diff
@@ -43,13 +43,13 @@ class RangeGen(axisName: String, from: Int, upto: Int, hop: Int) extends JavaGen
 }
 
 
-class EnumerationGen[T](axisName: String, xs: Array[T]) extends JavaGenerator[T] {
+class EnumerationGen[T](axisName: String, xs: Array[T], manifest: Manifest[T]) extends JavaGenerator[T] {
```
Comment (axel22) [inline src/main/scala/org/scalameter/japi/GenFactory.scala]: Same here.

File: src/main/scala/org/scalameter/japi/JavaPerformanceTest.scala [hunk 13]
```diff
@@ -210,7 +210,7 @@ abstract class Microbenchmark extends JavaPerformanceTest {
 abstract class HTMLReport extends JavaPerformanceTest {
   import Executor.Measurer
   import reporting._
-  def javaPersistor: org.scalameter.japi.Persistor = new org.scalameter.japi.SerializationPersistor
+  def javaPersistor: org.scalameter.japi.Persistor = new org.scalameter.japi.GZIPJSONSerializationPersistor
```
Comment (axel22) [inline src/main/scala/org/scalameter/japi/JavaPerformanceTest.scala]: See earlier comment.

File: src/main/scala/org/scalameter/persistence/IOStreamPersistor.scala [hunk 14]
```diff
@@ -0,0 +1,60 @@
+package org.scalameter
+package persistence
+
+import java.io._
+import Key.reports._
+
+
+trait IOStreamPersistor[In <: InputStream, Out <: OutputStream] extends Persistor {
```
Comment (axel22) [inline src/main/scala/org/scalameter/persistence/IOStreamPersistor.scala]: Please add some comments.

File: src/main/scala/org/scalameter/persistence/SerializationPersistor.scala [hunk 15]
```diff
@@ -3,62 +3,28 @@ package persistence
...
 
-import java.util.Date
 import java.io._
-import collection._
 import Key.reports._
 
 
 
-case class SerializationPersistor(path: File) extends Persistor {
+case class SerializationPersistor(path: File) extends IOStreamPersistor[ObjectInputStream, ObjectOutputStream] {
```
Comment (axel22) [inline src/main/scala/org/scalameter/persistence/SerializationPersistor.scala]: I like this! :)

Comment (axel22) [inline src/main/scala/org/scalameter/persistence/SerializationPersistor.scala]: Please add a unit test for this refactoring.

**Comment by axel22** [comment]
One more general comment - it would be helpful if you removed some of the non-essential functionality away from the PR, and refactor it into separate, smaller PRs.
The main comment I had (refactoring into `Pickler[T]`) can be addressed in this PR.

File: src/test/java/org/scalameter/examples/JavaRegressionTest3.java [hunk 16]
```diff
@@ -21,7 +22,7 @@ public Persistor javaPersistor() {
 			  .put("exec.outliers.covMultiplier", 1.5)
 			  .put("exec.outliers.suspectPercent", 40);
 			public JavaGenerator<LinkedList<Integer>> generator() {
-				JavaGenerator<Integer> sizes = new SingleGen("size", 5000000);
+				JavaGenerator<Integer> sizes = new SingleGen("size", 5000000, ManifestFactory$.MODULE$.Int());
```
Comment (axel22) [inline src/test/java/org/scalameter/examples/JavaRegressionTest3.java]: I would suggest improving `japi` to have custom `*SingleGen` versions for `Int` and `Double`.

File: src/test/scala/org/scalameter/persistence/PersistorTest.scala [hunk 17]
```diff
@@ -0,0 +1,35 @@
+package org.scalameter.persistence
+
+import org.scalameter._
+import org.scalatest.Matchers
+
+
+trait PersistorTest { this: Matchers =>
```
Comment (axel22) [inline src/test/scala/org/scalameter/persistence/PersistorTest.scala]: Nice!.

**Comment by axel22** [comment]
Another comment: please validate the quality of the persistor - compare the data output size between the `SerializationPersistor` and the new one. Test and compare the running times for large datasets. Do both with and without compression.

File: project/Build.scala (hunk 5 above)
Comment (kjanosz) [inline project/Build.scala]: Ouch, sorry for that. This is residue after first design (unpushed), where specific case classes were annotated to use specific deserializers instead of current design.

File: src/main/scala/org/scalameter/persistence/SerializationPersistor.scala (hunk 15 above)
Comment (kjanosz) [inline src/main/scala/org/scalameter/persistence/SerializationPersistor.scala]: Could you please elaborate more on that - e.g. what kind of unit test do you expect?
I'm asking because there is already a unit test ([SerializationPersistorTest](https://github.com/kjanosz/scalameter/blob/json-serialization-persistor/src/test/scala/org/scalameter/persistence/SerializationPersistorTest.scala)) along others *PersistorTest's that check if given persistor correctly serializes and deserializes output.

**Comment by kjanosz** [comment]
> One more general comment - it would be helpful if you removed some of the non-essential functionality away from the PR, and refactor it into separate, smaller PRs.
The main comment I had (refactoring into Pickler[T]) can be addressed in this PR.

Since refactoring into Pickler[T] is the first step that does not use any other code written here, so maybe better idea would be to adress it in different PR? - it would be definitely more convenient for me ;-)
I would split this PR into 3/4 PRs (Pickler[T] and custom jackson codecs both in separate PRs, then actual persistors in this PR and tests in separate PR?) - does this look OK to you? 

> Another comment: please validate the quality of the persistor - compare the data output size between the SerializationPersistor and the new one. Test and compare the running times for large datasets. Do both with and without compression.

I guess that comparing output size should be done in some unit test.
You also mention about testing it for large dataset - should I prepare unit test for that case or testing it on some external benchmark, like [this one you mentioned earlier](https://github.com/scala-blitz/scala-blitz/blob/master/src/test/scala/org/scala/optimized/test/scalameter/benchmarks.scala) and reporting result to you by e.g. email would be sufficient?

File: scalameter-core/src/main/scala/org/scalameter/Parameter.scala (hunk 3 above)
Comment (kjanosz) [inline scalameter-core/src/main/scala/org/scalameter/Parameter.scala]: I was hoping too that json serialization can be handled without any custom code. But most scala libraries are typeclass-based so they do not like this `Any` part in maps. I've also checked `genson` and given a try `scala-pickling` but only with `Jackson` I was able to made a solution within assumed time that partially works.

I agree that deserializing `Manifest` is complex and in fact I don't like it :-(
I like your idea with `Pickler[T]`, although one thing that is not clear to me is with case when:

- someone defines generator that needs custom `Pickler`
- (s)he does not want to write custom `Pickler`, but just to use `SerializationPersistor`  

It would throw compilation error (it won't find any suitable `Pickler`), am I right?
So, should I also write stub like:
```
implicit object AnyPickler extends Pickler[Any] {
  def pickle(x: T): Array[Byte]  = sys.error("Pickling for unknown type is disallowed. Please use SerializationPersistor instead.")
  def unpickle(a: Array[Byte]): T
}
```
this one, that will throw error in runtime when trying to use it with json based persistors?
Or just it should up to the user (writing some stub) in this case?

As for `Key`, I think they can be done as they are in this PR. `Manifest` is not persisted - only full path of `Key` is stored, e.g. for `scope` path `dsl.scope` is saved and based only on this `Key` in reconstructed using `Key.parseKey`. 
Or do you think it's a bad idea?
Also, I've chosen `Manifest` and not e.g. `TypeTag` only because:

- there is already conversion function from `Manifest` to Jackson's `JavaType` in `jackson-scala` module
- as far as I know `TypeTag`s are not thread safe in [`Scala 2.10`](https://issues.scala-lang.org/browse/SI-6240?focusedCommentId=71707&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-71707)

File: src/main/scala/org/scalameter/persistence/SerializationPersistor.scala (hunk 15 above)
Comment (axel22) [inline src/main/scala/org/scalameter/persistence/SerializationPersistor.scala]: Stale comment, my bad - I only saw it later, and then forgot about my comment.

**Comment by axel22** [comment]
Sure, I'm fine if you separate this into several PRs, as you see fit. The more incremental the better - what you propose sounds good.

About comparing the output size, I guess having some sort of a unit test that writes the serialization data to a byte stream/array, and compares the output sizes of `Serialization*` and new `Persistor`s with some constant is ssufficient.

**Comment by axel22** [comment]
For the bigger benchmark, you can do a one-off as a sanity check, and we can discuss the results.

File: scalameter-core/src/main/scala/org/scalameter/Parameter.scala (hunk 3 above)
Comment (axel22) [inline scalameter-core/src/main/scala/org/scalameter/Parameter.scala]: > they define generator that needs custom `Pickler`

They better implement it themselves in this case, and maintain the `Pickler` for their custom data type. This custom `Pickler` should be available on the classpath when deserializing (since picklers would be instantiated via reflection), and the custom generator and test author should ensure that.

This is rare anyway, since they will have trouble visualizing the results in the `HTMLReporter`.

> they do not want to write custom `Pickler`, but just to use `SerializationPersistor`

For existing factory methods such as `single` and `enumerate`: they can then use the explicitly specified/imported `errorPickler`, which just throws an error when used. `SerializationPersistor` won't invoke picklers.

For custom `Gen` implementations: they will have to provide the `Pickler` in the `Gen`'s `Parameters` object. They can reuse the `errorPickler` for this.

Comment (axel22) [inline scalameter-core/src/main/scala/org/scalameter/Parameter.scala]: I'd call `AnyPickler` `ErrorPickler` instead.

Comment (axel22) [inline scalameter-core/src/main/scala/org/scalameter/Parameter.scala]: Yes, unless they import the `errorPickler` (alternative name `noPickler: NoPickler`), for such data types creating a `Gen` is a compilation error. When they specify it, it is a runtime error should they decide to use any pickler-based persistor.

Comment (axel22) [inline scalameter-core/src/main/scala/org/scalameter/Parameter.scala]: Since keys and parameters are kind of similar concepts, why not piggy-back one on top of the other - could you use `Pickler`s to serialize/deserialize values under specific keys? We can similarly limit the `Key`s to a fixed set of data types to reduce complexity.

Comment (kjanosz) [inline scalameter-core/src/main/scala/org/scalameter/Parameter.scala]: > Since keys and parameters are kind of similar concepts, why not piggy-back one on top of the other - could you use Picklers to serialize/deserialize values under specific keys?

I will do that and it will definitely be more coherent. 
I was just asking because I don't see much reduced complexity.
Moreover, current proposed implementation (with `def typeHint: Manifest[T]` in `Key[T]`) is more space compact, because there is no need to save any class name along with key and value. 
Guess I didn't catch something.

Comment (axel22) [inline scalameter-core/src/main/scala/org/scalameter/Parameter.scala]: Besides coherency, the advantage is that everybody who implements another `Pickler`-based serializer, only needs to care about a transformation from `Array[Byte]` to serializer's data format, instead of any `Key` type `T` to the serializer's data format.
About space consumption, you could have a lookup table `Int -> String` for class names at the beginning of the serialized file.

I don't think this gain is huge, though, since:
- we have only a few data-types we care about serializing
- new serializers are not implemented often

So, I Iet you decide what's most convenient. If you decide to keep the current implementation, make sure that you move the `scalaMeterModule.scala` file to the basic project.
